### PR TITLE
feat: add Mutiny support

### PR DESCRIPTION
**Pull request by aoudiamoncef**
Hi @martinbonnin 

I'm back with [Mutiny](https://smallrye.io/smallrye-mutiny) integration for Apollo client. 

Thanks to project maintainers @cescoffier && @jponge whose gave me a lot of feedbacks. 

See our [discussion](https://github.com/smallrye/smallrye-mutiny/discussions/603).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,7 +21,7 @@ This library is designed primarily with Android in mind, but you can use it in a
 * Normalized cache
 * File uploads
 * Custom scalar types
-* Support for RxJava2, RxJava3, Coroutines and Reactor
+* Support for RxJava2, RxJava3, Coroutines, Reactor and Mutiny
```
Comment (martinbonnin) [inline README.md:24]:
```suggestion
* Reactive bindings for: RxJava2, RxJava3, Coroutines, Reactor and Mutiny
```

File: docs/source/advanced/mutiny.mdx [hunk 2]
```diff
@@ -0,0 +1,87 @@
+---
+title: Mutiny
+---
```
Comment (martinbonnin) [inline docs/source/advanced/mutiny.mdx:3]: Add this to `docs/gatsby-config.js` as well?

**Comment by martinbonnin** [comment]
@aoudiamoncef CI fails with 

```
incompatible types: io.smallrye.mutiny.Uni<java.lang.Object> cannot be converted to io.smallrye.mutiny.Uni<com.apollographql.apollo.api.Response<T>>
```

Can you take a look?

File: apollo-mutiny-support/src/main/java/com/apollographql/apollo/mutiny/MutinyApollo.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                            emitter.complete(response);
+                          }
+
+                          @Override public void onFailure(@NotNull ApolloException e) {
+                            emitter.fail(e);
+                          }
+
+                          @Override public void onStatusEvent(@NotNull ApolloCall.StatusEvent event) {
+                            if (event == ApolloCall.StatusEvent.COMPLETED) {
+                              emitter.complete(null);
```
Comment (cescoffier) [inline apollo-mutiny-support/src/main/java/com/apollographql/apollo/mutiny/MutinyApollo.java:79]: If the subscriber already got the response, this call will be dropped. I believe it's ok but better double-check.

File: docs/source/advanced/mutiny.mdx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+Java:
+```java
+// Create a query object
+EpisodeHeroName query = EpisodeHeroName.builder().episode(Episode.EMPIRE).build();
+
+// Create an ApolloCall object
+ApolloCall<EpisodeHeroName.Data> apolloCall = apolloClient.query(query);
+
+// Mutiny Unit
+Unit<Response<EpisodeHeroName.Data>> uni = MutinyApollo.from(apolloCall);
```
Comment (phillip-kruger) [inline docs/source/advanced/mutiny.mdx:45]: Should this maybe be `Uni` (rather that `Unit`) ?

File: docs/source/advanced/mutiny.mdx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+Java:
+```java
+//Create a query object
+EpisodeHeroName query = EpisodeHeroName.builder().episode(Episode.EMPIRE).build();
+
+//Create an ApolloPrefetch object
+ApolloPrefetch<EpisodeHeroName.Data> apolloPrefetch = apolloClient.prefetch(query);
+
+//Mutiny Unit<Void>
+Unit<Void> uni = MutinyApollo.from(apolloPrefetch);
```
Comment (phillip-kruger) [inline docs/source/advanced/mutiny.mdx:69]: Same here (Except if Unit is an internal class ?)

File: docs/source/advanced/mutiny.mdx [hunk 6]
```diff
@@ -0,0 +1,87 @@
+---
+title: Mutiny
+---
+
+Apollo Android includes support for Mutiny.
+
+Apollo types can be converted to Mutiny `Unit` *types* using wrapper functions `MutinyApollo.from(...)` in Java or using extension functions in Kotlin.
```
Comment (phillip-kruger) [inline docs/source/advanced/mutiny.mdx:7]: Unit - > Uni ?

File: docs/source/advanced/mutiny.mdx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+Conversion is done according to the following table:
+
+| Apollo type |  Mutiny type|
+| :--- | :--- |
+| `ApolloCall<T>` | `Unit<Response<T>>` |
+| `ApolloSubscriptionCall<T>` | `Multi<Response<T>>` |
+| `ApolloQueryWatcher<T>` | `Unit<Response<T>>` |
+| `ApolloStoreOperation<T>` | `Unit<T>` |
+| `ApolloPrefetch` | `Unit<Void>` |
```
Comment (phillip-kruger) [inline docs/source/advanced/mutiny.mdx:17]: Unit -> Uni

File: docs/source/advanced/mutiny.mdx [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+Add the following `dependency`:
+
+```groovy
+// Mutiny support
+implementation 'com.apollographql.apollo:apollo-mutiny-support:x.y.z'
+```
+
+## Usage examples
+
+### Converting `ApolloCall` to an `Unit`
```
Comment (phillip-kruger) [inline docs/source/advanced/mutiny.mdx:34]: Unit -> Uni

File: docs/source/advanced/mutiny.mdx (hunk 6 above)
Comment (aoudiamoncef) [reply docs/source/advanced/mutiny.mdx:7]: CTRL + R you know ): the consequences. I'll fix this

File: apollo-mutiny-support/src/main/java/com/apollographql/apollo/mutiny/MutinyApollo.java (hunk 3 above)
Comment (aoudiamoncef) [reply apollo-mutiny-support/src/main/java/com/apollographql/apollo/mutiny/MutinyApollo.java:79]: In Rxjava we have: 

```java
          if (event == ApolloCall.StatusEvent.COMPLETED && !emitter.isDisposed()) {
            emitter.onComplete();
          }
```
I don't know how to achieve equivalent check in Mutiny.

**Comment by martinbonnin** [comment]
And now complaining on metalava signatures. @aoudiamoncef , should I update them ?

**Comment by aoudiamoncef** [comment]
> And now complaining on metalava signatures. @aoudiamoncef , should I update them ?

Fixed, it was an issue with package name of `io.smallrye.mutiny.subscription.BackPressureStrategy`

File: apollo-mutiny-support/src/main/java/com/apollographql/apollo/mutiny/MutinyApollo.java (hunk 3 above)
Comment (jponge) [reply apollo-mutiny-support/src/main/java/com/apollographql/apollo/mutiny/MutinyApollo.java:79]: You may register a callback for _terminatIon_: https://javadoc.io/static/io.smallrye.reactive/mutiny/0.18.1/io/smallrye/mutiny/subscription/UniEmitter.html

Comment (jponge) [reply apollo-mutiny-support/src/main/java/com/apollographql/apollo/mutiny/MutinyApollo.java:79]: Other than that agreed with @cescoffier

**Comment by jponge** [comment]
🥳

**Pull request merged**